source/h1_encoder.c

```diff
@@ -712,116 +443,308 @@ int aws_chunk_line_from_options(struct aws_http1_chunk_options *options, struct
-            return AWS_OP_SUCCESS;
-        }
+/* Select next chunk to work on.
+ * Encoder is essentially "paused" here if no chunks are available. */
+static int s_state_fn_chunk_next(struct aws_h1_encoder *encoder, struct aws_byte_buf *dst) {
+    (void)dst;
+
+    if (aws_linked_list_empty(&encoder->message->chunk_list)) {
+        /* Remain in this state until more chunks arrive */
+        ENCODER_LOG(TRACE, encoder, "No chunks ready to send, waiting for more...");
+        return AWS_OP_SUCCESS;
+    }
+
+    /* Set next chunk and go to next state */
+    struct aws_linked_list_node *node = aws_linked_list_front(&encoder->message->chunk_list);
+    encoder->current_chunk = AWS_CONTAINER_OF(node, struct aws_h1_chunk, node);
+    encoder->chunk_count++;
+    ENCODER_LOGF(
+        TRACE,
+        encoder,
+        "Begin sending chunk %zu with size %" PRIu64,
+        encoder->chunk_count,
+        encoder->current_chunk->data_size);
+
+    return s_switch_state(encoder, AWS_H1_ENCODER_STATE_CHUNK_LINE);
+}
 
-        /* Copy data from outgoing_head_buf */
-        struct aws_byte_buf *src = &encoder->message->outgoing_head_buf;
-        size_t src_progress = (size_t)encoder->progress_bytes;
-        size_t src_remaining = src->len - src_progress;
-        size_t transferring = src_remaining < dst_available ? src_remaining : dst_available;
+/* Write out "chunk-size [chunk-ext] CRLF".
+ * This data is pre-encoded in the chunk's chunk_line buffer */
+static int s_state_fn_chunk_line(struct aws_h1_encoder *encoder, struct aws_byte_buf *dst) {
+    bool done = s_encode_buf(encoder, dst, &encoder->current_chunk->chunk_line);
+    if (!done) {
+        /* Remain in state until done writing line */
+        return AWS_OP_SUCCESS;
+    }
 
-        bool success = aws_byte_buf_write(dst, src->buffer + src_progress, transferring);
-        (void)success;
-        AWS_ASSERT(success);
+    /* Pick next state */
+    if (encoder->current_chunk->data_size == 0) {
+        /* If data_size is 0, then this was the last chunk, which has no body.
+         * Mark it complete and move on to trailer. */
+        ENCODER_LOG(TRACE, encoder, "Final chunk complete");
+        s_clean_up_current_chunk(encoder);
 
-        encoder->progress_bytes += transferring;
+        return s_switch_state(encoder, AWS_H1_ENCODER_STATE_CHUNK_TRAILER);
+    }
 
-        ENCODER_LOGF(
-            TRACE,
-            encoder,
-            "Writing to message, outgoing head progress %" PRIu64 "/%zu.",
-            encoder->progress_bytes,
-            encoder->message->outgoing_head_buf.len);
+    return s_switch_state(encoder, AWS_H1_ENCODER_STATE_CHUNK_BODY);
+}
+
+/* Write out data for current chunk */
+static int s_state_fn_chunk_body(struct aws_h1_encoder *encoder, struct aws_byte_buf *dst) {
+    bool done;
+    if (s_encode_stream(encoder, dst, encoder->message->body, encoder->message->content_length, &done)) {
```

Comment:
I haven't seen where we update the `encoder->message->content_length` to `encoder->message->body_chunks->current_chunk->data_size`. We may need to update it? Or should it be `s_encode_stream(encoder, dst, encoder->message->body, encoder->message->body_chunks->current_chunk->data_size, &done)` ?

Reply:
Oh, the `encoder->message->body` should be `encoder->current_chunk->data`?

Reply:
yup 🤦‍♂️